### PR TITLE
`ActiveJob::QueueAdapters::*` are no longer singletons

### PR DESCRIPTION
**Pull request by tamird**
@matthewd @jeremy

### CODE REVIEW

File: activejob/lib/active_job/queue_adapter.rb [hunk 1]
```diff
@@ -14,19 +14,23 @@ module ClassMethods
...
       # information.
-      def queue_adapter=(name_or_adapter)
+      def queue_adapter=(name_or_adapter_or_class)
         @@queue_adapter = \
-          case name_or_adapter
+          case name_or_adapter_or_class
           when Symbol, String
-            load_adapter(name_or_adapter)
+            load_adapter(name_or_adapter_or_class)
           else
```
Comment (cristianbica) [inline activejob/lib/active_job/queue_adapter.rb]: how about?
```ruby
when Class
  name_or_adapter_or_class.new
else
  name_or_adapter_or_class
```

Comment (tamird) [inline activejob/lib/active_job/queue_adapter.rb]: had to do something a bit more complex to provide the desired validation

Comment (jeremy) [inline activejob/lib/active_job/queue_adapter.rb]: The `Class` instance may well be the adapter singleton, so we have to duck-type.

File: activejob/lib/active_job/queue_adapter.rb [hunk 2]
```diff
@@ -14,20 +14,26 @@ module ClassMethods
...
-          when Symbol, String
-            load_adapter(name_or_adapter)
+      def queue_adapter=(name_or_adapter_or_class)
+        case name_or_adapter_or_class
+        when Symbol, String
+          self.queue_adapter = load_adapter_class(name_or_adapter_or_class)
+        else
+          @@queue_adapter = if name_or_adapter_or_class.respond_to?(:enqueue)
+            name_or_adapter_or_class
+          elsif name_or_adapter_or_class.is_a?(Class) && name_or_adapter_or_class.public_method_defined?(:enqueue)
```
Comment (jeremy) [inline activejob/lib/active_job/queue_adapter.rb]: Think this check should be in `load_adapter`: "give me an adapter name and I'll return an instance"

Comment (jeremy) [inline activejob/lib/active_job/queue_adapter.rb]: Nevermind. We still have to support `.queue_adapter = NameOfExistingAdapter` eh.

Comment (tamird) [inline activejob/lib/active_job/queue_adapter.rb]: I don't feel strongly - will do

Comment (jeremy) [inline activejob/lib/active_job/queue_adapter.rb]: (Good spot to mark for deprecation, though)

Comment (tamird) [inline activejob/lib/active_job/queue_adapter.rb]: actually, this check is doing something else - by now we know we didn't get a name, we either got an instance or a class, and we need to call `new` or not accordingly - the name case is handled in the previous branch

File: activejob/lib/active_job/queue_adapter.rb [hunk 3]
```diff
@@ -14,20 +14,27 @@ module ClassMethods
...
-            load_adapter(name_or_adapter)
+      def queue_adapter=(name_or_adapter_or_class)
+        case name_or_adapter_or_class
+        when Symbol, String
+          self.queue_adapter = load_adapter(name_or_adapter_or_class)
+        else
+          @@queue_adapter = if name_or_adapter_or_class.respond_to?(:enqueue)
+            name_or_adapter_or_class
+          elsif name_or_adapter_or_class.is_a?(Class) && name_or_adapter_or_class.public_method_defined?(:enqueue)
+            ActiveSupport::Deprecation.warn 'Passing a Class is deprecated. Please pass an instance instead.'
```
Comment (jeremy) [inline activejob/lib/active_job/queue_adapter.rb]: Sketching, clarifying what class/instance mean to people who find themselves in this situation:

Passing an adapter class is deprecated and will be removed in Rails 5.1. Please pass an adapter name (`.queue_adapter = :#{name_or_adapter_or_class.name.demodulize.remove('Adapter').underscore}`) or an instance (`queue_adapter = #{name_or_adapter_or_class.name}.new`) instead.

File: activejob/lib/active_job/queue_adapter.rb [hunk 4]
```diff
@@ -14,20 +14,27 @@ module ClassMethods
...
+            name_or_adapter_or_class.new
           else
-            name_or_adapter if name_or_adapter.respond_to?(:enqueue)
+            raise ArgumentError
```
Comment (jeremy) [inline activejob/lib/active_job/queue_adapter.rb]: Could flip the main `if` clause down here and skip the `.respond_to? :enqueue` check. Expect compliance with duck type rather than enforcing it.

File: activejob/test/cases/logging_test.rb [hunk 5]
```diff
@@ -6,7 +6,7 @@
 require 'jobs/nested_job'
 require 'models/person'
 
-class AdapterTest < ActiveSupport::TestCase
+class LoggingTest < ActiveSupport::TestCase
```
Comment (jeremy) [inline activejob/test/cases/logging_test.rb]: ohai

File: activejob/test/cases/queue_adapter_test.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
+require 'helper'
+
+class QueueAdapterTest < ActiveJob::TestCase
+  test 'should forbid nonsense arguments' do
+    assert_raises(ArgumentError) { ActiveJob::Base.queue_adapter = Mutex }
+    assert_raises(ArgumentError) { ActiveJob::Base.queue_adapter = Mutex.new }
+  end
```
Comment (jeremy) [inline activejob/test/cases/queue_adapter_test.rb]: Think these should be accepted. They'll fail when used, as duck types should.

File: activejob/test/cases/test_case_test.rb [hunk 7]
```diff
@@ -9,6 +9,6 @@ def test_include_helper
   end
 
   def test_set_test_adapter
-    assert_equal ActiveJob::QueueAdapters::TestAdapter, self.queue_adapter
+    assert_equal ActiveJob::QueueAdapters::TestAdapter, self.queue_adapter.class
```
Comment (jeremy) [inline activejob/test/cases/test_case_test.rb]: Can use `assert_kind_of`

File: activejob/test/support/integration/test_case_helpers.rb [hunk 8]
```diff
@@ -28,7 +28,7 @@ def clear_jobs
     end
 
     def adapter_is?(adapter)
```
Comment (jeremy) [inline activejob/test/support/integration/test_case_helpers.rb]: Would rename the argument to `adapter_class` to reflect the shift in usage

Comment (tamird) [inline activejob/test/support/integration/test_case_helpers.rb]: addressed

File: activejob/test/cases/test_case_test.rb (hunk 7 above)
Comment (tamird) [inline activejob/test/cases/test_case_test.rb]: addressed

File: activejob/test/cases/queue_adapter_test.rb (hunk 6 above)
Comment (tamird) [inline activejob/test/cases/queue_adapter_test.rb]: as discussed, leaving as is

File: activejob/test/cases/logging_test.rb (hunk 5 above)
Comment (tamird) [inline activejob/test/cases/logging_test.rb]: hi :) i put this change in its own commit

File: activejob/lib/active_job/queue_adapter.rb (hunk 4 above)
Comment (tamird) [inline activejob/lib/active_job/queue_adapter.rb]: as discussed, leaving as is

File: activejob/lib/active_job/queue_adapter.rb (hunk 3 above)
Comment (tamird) [inline activejob/lib/active_job/queue_adapter.rb]: done

**Comment by tamird** [comment]
https://travis-ci.org/rails/rails/builds/54024576

**Comment by cristianbica** [comment]
Nice ❤️
